File: Sources/COCHelperCore/VillageDetailFlatRowCache.swift
import Foundation

/// Issue #212：村庄详情扁平 row 元数据缓存。
///
/// 与 `VillageProjectionCache` 的 render 身份对齐；筛选/排序/分类变化自然
/// miss。`.remaining` 排序依赖 `now`，tick 间不缓存。View body 仍由
/// LazyVStack 按需构建；本缓存只避免重复分配 row descriptor 数组。
public final class VillageDetailFlatRowCache {
    /// 与投影缓存对齐的静态输入身份（不含 now）。
    public struct RenderIdentityKey: Hashable, Sendable {
        public let villageID: UUID
        public let villageName: String
        public let snapshotFingerprint: String
        public let base: TrackerBase
        public let manualFingerprint: String?
        public let catalogEpoch: Int
        public let catalogVersion: String?
        public let phaseBucket: PhaseBucket

        public init(
            villageID: UUID,
            villageName: String,
            snapshotFingerprint: String,
            base: TrackerBase,
            manualFingerprint: String?,
            catalogEpoch: Int,
            catalogVersion: String?,
            phaseBucket: PhaseBucket
        ) {
            self.villageID = villageID
            self.villageName = villageName
            self.snapshotFingerprint = snapshotFingerprint
            self.base = base
            self.manualFingerprint = manualFingerprint
            self.catalogEpoch = catalogEpoch
            self.catalogVersion = catalogVersion
            self.phaseBucket = phaseBucket
        }

        public init?(
            village: VillageProfile,
            base: TrackerBase,
            now: Date,
            manualUpgradeCore: ManualUpgradeCore?,
            catalogEpoch: Int,
            catalog: GameCatalog?,
            seasonalPhases: SeasonalPhaseTable
        ) {
            guard let snapshot = village.accountSnapshot else { return nil }
            self.villageID = village.id
            self.villageName = village.name
            self.snapshotFingerprint = snapshot.contentFingerprint
            self.base = base
            self.manualFingerprint = manualUpgradeCore?.contentFingerprint
            self.catalogEpoch = catalogEpoch
            self.catalogVersion = catalog?.gameVersion
            self.phaseBucket = seasonalPhases.bucket(at: now)
        }
    }

    /// 展示筛选身份（search/state/sort/category；不含 now）。
    public struct FilterKey: Hashable, Sendable {
        public let searchText: String
        public let stateFilter: UpgradeDisplayStateFilter?
        public let sortOrder: UpgradeDisplaySort
        /// 分类 chip 的稳定编码（由 UI 层提供，如 `all` / `display:walls`）。
        public let categoryFilterKey: String

        public init(
            searchText: String,
            stateFilter: UpgradeDisplayStateFilter?,
            sortOrder: UpgradeDisplaySort,
            categoryFilterKey: String
        ) {
            self.searchText = searchText
            self.stateFilter = stateFilter
            self.sortOrder = sortOrder
            self.categoryFilterKey = categoryFilterKey
        }
    }

    private struct Entry {
        let rows: [VillageDetailFlatRow]
        let groupByInstanceID: [String: BuildingGroup]
    }

    private struct Key: Hashable {
        let render: RenderIdentityKey
        let filter: FilterKey
    }

    private var entry: Entry?
    private var entryKey: Key?

    /// 构建次数（测试断言用）。
    public private(set) var buildCount = 0
    /// 命中次数（测试断言用）。
    public private(set) var hitCount = 0

    public init() {}

    public func removeAll() {
        entry = nil
        entryKey = nil
    }

    /// 取扁平 row 元数据；`sortDependsOnNow == true` 时跳过缓存（如 `.remaining`）。
    public func rows(
        renderKey: RenderIdentityKey,
        filterKey: FilterKey,
        sortDependsOnNow: Bool,
        displayGroups: [VillageDetailGroup],
        statsByKey: [String: VillageCategoryCompletion],
        buildingGroups: [BuildingGroup]
    ) -> (rows: [VillageDetailFlatRow], groupByInstanceID: [String: BuildingGroup]) {
        let key = Key(render: renderKey, filter: filterKey)
        if !sortDependsOnNow, let entry, entryKey == key {
            hitCount += 1
            return (entry.rows, entry.groupByInstanceID)
        }

        buildCount += 1
        let groupByInstanceID = VillageDetailFlatRowProjection.groupByInstanceID(
            from: buildingGroups
        )
        let built = VillageDetailFlatRowProjection.build(
            displayGroups: displayGroups,
            statsByKey: statsByKey,
            groupByInstanceID: groupByInstanceID
        )
        let newEntry = Entry(rows: built, groupByInstanceID: groupByInstanceID)
        entry = newEntry
        entryKey = key
        return (newEntry.rows, newEntry.groupByInstanceID)
    }
}

File: packages/domain/src/village/village-detail-flat-row-cache.ts
import type { GameCatalog } from '../catalog/game-catalog';
import type { SeasonalPhaseTable, PhaseBucket } from '../catalog/seasonal-phase';
import type { ManualUpgradeCore } from '../manual/types';
import type { BuildingGroup } from './building-group-projection';
import type { TrackerBase } from './tracker';
import type { VillageProfile } from '../import/types';
import {
  buildVillageDetailFlatRows,
  groupBuildingGroupsByInstanceID,
  type VillageDetailFlatRow,
} from './village-detail-flat-row';
import type { VillageCategoryCompletion, VillageDetailGroup } from './village-detail-projection';
import type { VillageProjectionCacheRenderResult } from './village-projection-cache';

export type UpgradeDisplayStateFilter =
  'available' | 'manualActive' | 'importedActive' | 'completed' | 'needsReimport' | 'unknown';

export type UpgradeDisplaySort =
  'remaining' | 'categoryName' | 'level' | 'stageMax' | 'recentlyChanged';

export type VillageDetailFlatRowRenderIdentityKey = {
  readonly villageID: string;
  readonly villageName: string;
  readonly snapshotFingerprint: string;
  readonly base: TrackerBase;
  readonly manualFingerprint: string | null;
  readonly catalogEpoch: number;
  readonly catalogVersion: string | null;
  readonly phaseBucket: PhaseBucket;
  readonly projectionGeneration: bigint;
};

export type VillageDetailFlatRowFilterKey = {
  readonly searchText: string;
  readonly stateFilter: UpgradeDisplayStateFilter | null;
  readonly sortOrder: UpgradeDisplaySort;
  readonly categoryFilterKey: string;
};

type CacheEntry = {
  readonly rows: readonly VillageDetailFlatRow[];
  readonly groupByInstanceID: Record<string, BuildingGroup>;
};

type CacheKey = {
  readonly render: VillageDetailFlatRowRenderIdentityKey;
  readonly filter: VillageDetailFlatRowFilterKey;
};

export function createVillageDetailFlatRowRenderIdentityKey(input: {
  readonly village: VillageProfile;
  readonly render: VillageProjectionCacheRenderResult;
  readonly base: TrackerBase;
  readonly nowMs: number;
  readonly manualUpgradeCore: ManualUpgradeCore | null | undefined;
  readonly catalogEpoch: number;
  readonly catalog: GameCatalog | null | undefined;
  readonly seasonalPhases: SeasonalPhaseTable;
}): VillageDetailFlatRowRenderIdentityKey | null {
  const snapshot = input.village.accountSnapshot;
  if (snapshot === null || snapshot === undefined) {
    return null;
  }
  return {
    villageID: input.village.id,
    villageName: input.village.name,
    snapshotFingerprint: snapshot.contentFingerprint,
    base: input.base,
    manualFingerprint: input.manualUpgradeCore?.contentFingerprint ?? null,
    catalogEpoch: input.catalogEpoch,
    catalogVersion: input.catalog?.gameVersion ?? null,
    phaseBucket: input.seasonalPhases.bucket(input.nowMs),
    projectionGeneration: input.render.projectionGeneration,
  };
}

export class VillageDetailFlatRowCache {
  private entry: CacheEntry | null = null;
  private entryKey: CacheKey | null = null;
  private _buildCount = 0;
  private _hitCount = 0;

  get buildCount(): number {
    return this._buildCount;
  }

  get hitCount(): number {
    return this._hitCount;
  }

  removeAll(): void {
    this.entry = null;
    this.entryKey = null;
  }

  rows(input: {
    readonly renderKey: VillageDetailFlatRowRenderIdentityKey;
    readonly filterKey: VillageDetailFlatRowFilterKey;
    readonly sortDependsOnNow: boolean;
    readonly displayGroups: readonly VillageDetailGroup[];
    readonly statsByKey: Readonly<Record<string, VillageCategoryCompletion>>;
    readonly buildingGroups: readonly BuildingGroup[];
  }): {
    readonly rows: readonly VillageDetailFlatRow[];
    readonly groupByInstanceID: Record<string, BuildingGroup>;
  } {
    const key: CacheKey = {
      render: input.renderKey,
      filter: input.filterKey,
    };
    if (
      !input.sortDependsOnNow &&
      this.entry !== null &&
      this.entryKey !== null &&
      cacheKeysEqual(this.entryKey, key)
    ) {
      this._hitCount += 1;
      return {
        rows: this.entry.rows,
        groupByInstanceID: this.entry.groupByInstanceID,
      };
    }

    this._buildCount += 1;
    const groupByInstanceID = groupBuildingGroupsByInstanceID(input.buildingGroups);
    const built = buildVillageDetailFlatRows({
      displayGroups: input.displayGroups,
      statsByKey: input.statsByKey,
      groupByInstanceID,
    });
    const newEntry: CacheEntry = { rows: built, groupByInstanceID };
    this.entry = newEntry;
    this.entryKey = key;
    return {
      rows: newEntry.rows,
      groupByInstanceID: newEntry.groupByInstanceID,
    };
  }
}

function cacheKeysEqual(left: CacheKey, right: CacheKey): boolean {
  return (
    renderIdentityKeysEqual(left.render, right.render) &&
    left.filter.searchText === right.filter.searchText &&
    left.filter.stateFilter === right.filter.stateFilter &&
    left.filter.sortOrder === right.filter.sortOrder &&
    left.filter.categoryFilterKey === right.filter.categoryFilterKey
  );
}

function renderIdentityKeysEqual(
  left: VillageDetailFlatRowRenderIdentityKey,
  right: VillageDetailFlatRowRenderIdentityKey,
): boolean {
  return (
    left.villageID === right.villageID &&
    left.villageName === right.villageName &&
    left.snapshotFingerprint === right.snapshotFingerprint &&
    left.base === right.base &&
    left.manualFingerprint === right.manualFingerprint &&
    left.catalogEpoch === right.catalogEpoch &&
    left.catalogVersion === right.catalogVersion &&
    phaseBucketsEqual(left.phaseBucket, right.phaseBucket) &&
    left.projectionGeneration === right.projectionGeneration
  );
}

function phaseBucketsEqual(left: PhaseBucket, right: PhaseBucket): boolean {
  return (
    left.tableIdentity === right.tableIdentity &&
    left.startMs === right.startMs &&
    left.endMs === right.endMs
  );
}
